Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp

```diff
@@ -0,0 +1,544 @@
+    {
+        if (m_byte_order == ByteOrder::LittleEndian)
+            return TRY(m_stream->read_value<LittleEndian<T>>());
+        if (m_byte_order == ByteOrder::BigEndian)
+            return TRY(m_stream->read_value<BigEndian<T>>());
+        VERIFY_NOT_REACHED();
+    }
+
+    ErrorOr<void> read_next_idf_offset()
+    {
+        auto const next_block_position = TRY(read_value<u32>());
+
+        if (next_block_position != 0)
+            m_next_ifd = Optional<u32> { next_block_position };
+        else
+            m_next_ifd = OptionalNone {};
+        dbgln_if(TIFF_DEBUG, "Setting image file directory pointer to {}", m_next_ifd);
+        return {};
+    }
+
+    ErrorOr<void> read_image_file_header()
+    {
+        // Section 2: TIFF Structure - Image File Header
+
+        auto const byte_order = TRY(m_stream->read_value<u16>());
+
+        switch (byte_order) {
+        case 0x4949:
+            m_byte_order = ByteOrder::LittleEndian;
+            break;
+        case 0x4D4D:
+            m_byte_order = ByteOrder::BigEndian;
+            break;
+        default:
+            return Error::from_string_literal("TIFFImageDecoderPlugin: Invalid byte order");
+        }
+
+        auto const magic_number = TRY(read_value<u16>());
+
+        if (magic_number != 42)
+            return Error::from_string_literal("TIFFImageDecoderPlugin: Invalid magic number");
+
+        TRY(read_next_idf_offset());
+
+        return {};
+    }
+
+    ErrorOr<void> read_next_image_file_directory()
+    {
+        // Section 2: TIFF Structure - Image File Directory
+
+        if (!m_next_ifd.has_value())
+            return Error::from_string_literal("TIFFImageDecoderPlugin: Missing an Image File Directory");
+
+        TRY(m_stream->seek(m_next_ifd.value()));
+
+        auto const number_of_field = TRY(read_value<u16>());
+
+        for (u16 i = 0; i < number_of_field; ++i)
+            TRY(read_tag());
+
+        TRY(read_next_idf_offset());
+        return {};
+    }
+
+    ErrorOr<Type> read_type()
+    {
+        switch (TRY(read_value<u16>())) {
+        case to_underlying(Type::Byte):
+            return Type::Byte;
+        case to_underlying(Type::ASCII):
+            return Type::ASCII;
+        case to_underlying(Type::Short):
+            return Type::Short;
+        case to_underlying(Type::Long):
+            return Type::Long;
+        case to_underlying(Type::Rational):
+            return Type::Rational;
+        case to_underlying(Type::Undefined):
+            return Type::Undefined;
+        case to_underlying(Type::SLong):
+            return Type::SLong;
+        case to_underlying(Type::SRational):
+            return Type::SRational;
+        case to_underlying(Type::UTF8):
+            return Type::UTF8;
+        default:
+            return Error::from_string_literal("TIFFImageDecoderPlugin: Unknown type");
+        }
+    }
+
+    static constexpr u8 size_of_type(Type type)
+    {
+        switch (type) {
```

Comment:
Given we only call this with things that are known to be a valid Type and Type values being small consecutive integers, wdyt about doing 

```
constexpr u8 TiffTypeSizes[] = { -1, 1, 1, 2, 4, 8, 1, 1, 2, 4, 8 };
return TiffTypeSizes[(int)type];
```

?

(Optionally with `[.Type::Byte] = 1` style initializers if you want to be more explicit)

Reply:
I wasn't convinced, but I tried and liked it. I now consider you responsible for implementing it in GCC.
`sorry, unimplemented: non-trivial designated initializers not supported`